toolchain/lower/file_context.cpp

```diff
@@ -126,9 +126,18 @@ auto FileContext::BuildDICompileUnit(llvm::StringRef module_name,
                                       /*RV=*/0);
 }
 
-auto FileContext::GetGlobal(SemIR::InstId inst_id) -> llvm::Value* {
+auto FileContext::GetGlobal(SemIR::InstId inst_id,
+                            SemIR::SpecificId specific_id) -> llvm::Value* {
   auto inst = sem_ir().insts().Get(inst_id);
 
+  if (specific_id.has_value()) {
+    inst_id = sem_ir().constant_values().GetInstIdIfValid(
+        GetConstantValueInSpecific(sem_ir(), specific_id, inst_id));
+    CARBON_CHECK(inst_id.has_value(),
+                 "Expected to find an instruction id for a specific in the "
+                 "file context");
+  }
+
   auto const_id = sem_ir().constant_values().Get(inst_id);
```

Comment:
```suggestion
  auto const_id = GetConstantValueInSpecific(sem_ir(), specific_id, inst_id));
```

I think we can unconditionally call `GetConstantValueInSpecific` to simplify this a bit. It handles the case where the `specific_id` is `None`.

`inst` is only used to compute `inst.type_id()` below; that should instead use `context.insts().Get(const_inst_id).type_id()` so that it uses the type from the specific constant rather than the generic instruction.

Also, the TODO on line 185 is done, so can now be removed.

(It might be nice to add the `specific_id` to the "Missing value" fatal error too, since we'll hit that instead in the case where your CHECK would have failed, and it seems useful for it to say that we were looking for the value in a specific.)